sqlx-mysql/src/transaction.rs

```diff
@@ -14,12 +16,21 @@ pub struct MySqlTransactionManager;
 impl TransactionManager for MySqlTransactionManager {
     type Database = MySql;
 
-    fn begin(conn: &mut MySqlConnection) -> BoxFuture<'_, Result<(), Error>> {
+    fn begin<'conn>(
+        conn: &'conn mut MySqlConnection,
+        statement: Option<Cow<'static, str>>,
+    ) -> BoxFuture<'conn, Result<(), Error>> {
         Box::pin(async move {
             let depth = conn.inner.transaction_depth;
-
-            conn.execute(&*begin_ansi_transaction_sql(depth)).await?;
-            conn.inner.transaction_depth = depth + 1;
+            if statement.is_some() && depth > 0 {
+                return Err(Error::InvalidSavePointStatement);
+            }
+            let statement = statement.unwrap_or_else(|| begin_ansi_transaction_sql(depth));
```

Comment:
I have a feeling Clippy might warn on this (arguably) redundant `.is_some()` check before the `.unwrap_or_else()`.

It might be a bit clearer in intent just to use a plain old `match`:

```suggestion
            let statement = match statement {
                // custom `BEGIN` statements are not allowed if we're already in a transaction
                // (we need to issue a `SAVEPOINT` instead)
                Some(_) if depth > 0 => return Err(Error::InvalidSavePointStatement),
                Some(statement) => statement,
                None => begin_ansi_transaction_sql(depth),
            };
```